src/reducers/computeRenderedColumns.js

```diff
@@ -0,0 +1,232 @@
+ *
+ * @providesModule computeRenderedColumns
+ */
+
+'use strict';
+
+import clamp from 'lodash/clamp';
+import columnWidths from 'columnWidths';
+import scrollbarsVisibleSelector from 'scrollbarsVisible';
+import { updateColumnWidth, updateColumnGroupWidth } from 'updateColumnWidth';
+
+export default function computeRenderedColumns(state, columnAnchor) {
+  // clone state
+  const newState = Object.assign({}, state);
+  
+  // get buffer and viewport range
+  const columnRange = calculateRenderedColumnRange(newState, columnAnchor);
+  
+  // update the offsets and buffer mapping
+  computeRenderedColumnOffsets(newState, columnRange, newState.scrolling);
+
+  // scrollX might have changed due to change in columns and offsets
+  let scrollX = state.scrollX;
+  if (columnRange.firstViewportCol !== columnRange.endViewportCol) {
+    scrollX = newState.columnOffsets[columnRange.firstViewportCol] - newState.firstColumnOffset;
+  }
+  const { maxScrollX } = columnWidths(state);
+  newState.scrollX = clamp(scrollX, 0, maxScrollX);
+
+  return newState;
+}
+
+/**
+ * Determine the range of columns to render (buffer and viewport)
+ * The leading and trailing buffer is based on a fixed count,
+ * while the viewport columns are based on their width and the viewport width.
+ * We use the columnAnchor to determine what either the first or last column
+ * will be, as well as the offset.
+ *
+ * NOTE (jordan) This alters state so it shouldn't be called
+ * without state having been cloned first.
+ *
+ * @param {!Object} state
+ * @param {{
+ *   firstIndex: number,
+ *   firstOffset: number,
+ *   lastIndex: number,
+ * }} columnAnchor
+ * @return {{
+ *   endBufferCol: number,
+ *   endViewportCol: number,
+ *   firstBufferCol: number,
+ *   firstViewportCol: number,
+ * }}
+ * @private
+ */
+function calculateRenderedColumnRange(state, columnAnchor) {
+  const bufferColumnCount = 0; // TODO (pradeep): calculate this similar to bufferRowCount
+  const { availableScrollWidth, scrollableColumns } = columnWidths(state);
+  const columnCount = scrollableColumns.length;
+
+  if (availableScrollWidth === 0 || columnCount === 0) {
+    state.firstColumnIndex = 0;
+    state.endColumnIndex = 0;
+    state.firstColumnOffset = 0;
+    return {
+      endBufferCol: 0,
+      endViewportCol: 0,
+      firstBufferCol: 0,
+      firstViewportCol: 0,
+    };
+  }
+
+  // If our first or last index is greater than our columnCount,
+  // treat it as if the last column is at the end of the viewport
+  let { firstIndex, firstOffset, lastIndex } = columnAnchor;
+  if (firstIndex >= columnCount || lastIndex >= columnCount) {
+    lastIndex = columnCount - 1;
+  }
+
+  // Walk the viewport until filled with columns
+  // If lastIndex is set, walk backward so that column is the last in the viewport
+  let step = 1;
+  let startIdy = firstIndex;
+  let totalWidth = firstOffset;
+  if (lastIndex !== undefined) {
```

Comment:
if we scroll to end of table, `lastIndex` will be defined, hence we'll do the reverse way of iteration.

We start at the last column, and step backwards (hence the negative unit step)